crates/wasi-common/src/virtfs/pipe.rs

```diff
@@ -0,0 +1,412 @@
+//! than OS files.
+//!
+//! Some convenience constructors are included for common backing types like `Vec<u8>` and `String`,
+//! but the virtual pipes can be instantiated with any `Read` or `Write` type.
+//!
+//! Note that `poll_oneoff` is not supported for these types, so they do not match the behavior of
+//! real pipes exactly.
+use crate::handle::{Handle, HandleRights};
+use crate::wasi::{types, Errno, Result};
+use log::trace;
+use std::any::Any;
+use std::cell::{Cell, Ref, RefCell};
+use std::io::{self, Read, Write};
+use std::rc::Rc;
+
+/// A virtual pipe read end.
+///
+/// A variety of `From` impls are provided so that common pipe types are easy to create. For example:
+///
+/// ```
+/// # use wasi_common::WasiCtxBuilder;
+/// # use wasi_common::virtfs::pipe::ReadPipe;
+/// let mut ctx = WasiCtxBuilder::new();
+/// let stdin = ReadPipe::from("hello from stdin!");
+/// ctx.stdin(stdin);
+/// ```
+#[derive(Clone, Debug)]
+pub struct ReadPipe<R: Read + Any> {
+    rights: Cell<HandleRights>,
+    reader: Rc<RefCell<R>>,
+}
+
+impl<R: Read + Any> ReadPipe<R> {
+    /// Create a new pipe from a `Read` type.
+    ///
+    /// All `Handle` read operations delegate to reading from this underlying reader.
+    pub fn new(r: R) -> Self {
+        Self::from_shared(Rc::new(RefCell::new(r)))
+    }
+
+    /// Create a new pipe from a shareable `Read` type.
+    ///
+    /// All `Handle` read operations delegate to reading from this underlying reader.
+    pub fn from_shared(reader: Rc<RefCell<R>>) -> Self {
+        use types::Rights;
+        Self {
+            rights: Cell::new(HandleRights::new(
+                Rights::FD_DATASYNC
+                    | Rights::FD_FDSTAT_SET_FLAGS
+                    | Rights::FD_READ
+                    | Rights::FD_SYNC
+                    | Rights::FD_FILESTAT_GET
+                    | Rights::POLL_FD_READWRITE,
+                Rights::empty(),
+            )),
+            reader,
+        }
+    }
+
+    /// Try to convert this `ReadPipe<R>` back to the underlying `R` type.
+    ///
+    /// This will fail with `Err(self)` if multiple references to the underlying `R` exist.
+    pub fn try_into_inner(mut self) -> std::result::Result<R, Self> {
+        match Rc::try_unwrap(self.reader) {
+            Ok(rc) => Ok(RefCell::into_inner(rc)),
+            Err(reader) => {
+                self.reader = reader;
+                Err(self)
+            }
+        }
+    }
+}
+
+impl From<Vec<u8>> for ReadPipe<io::Cursor<Vec<u8>>> {
+    fn from(r: Vec<u8>) -> Self {
+        Self::new(io::Cursor::new(r))
+    }
+}
+
+impl From<&[u8]> for ReadPipe<io::Cursor<Vec<u8>>> {
+    fn from(r: &[u8]) -> Self {
+        Self::from(r.to_vec())
+    }
+}
+
+impl From<String> for ReadPipe<io::Cursor<String>> {
+    fn from(r: String) -> Self {
+        Self::new(io::Cursor::new(r))
+    }
+}
+
+impl From<&str> for ReadPipe<io::Cursor<String>> {
```

Comment:
Same idea about a `Cursor<&'a str>` here too, but I'm not as sure that this'll Just Work. Might be worth trying?